modules/schema_validation.lua

```diff
@@ -35,315 +26,264 @@ local function dump(o)
-      if (v1:attr("messagetype") == msg_type and v1:parent():attr("name")== class_) then
-        if (type(v1:children() == 'table')) then
-          for _, v2 in ipairs(v1:children()) do
-            local class_type = v2:attr('type')
-            if(type(v2:attr('name')) ~= 'nil') then
-              local tmp = {}
-              if ( types.classes[class_type]) then
-                tmp['class'] = string.format("%s",class_type)
-              end
-              if (types.enum[class_type]) then
-                tmp['class'] = "enum"
-                tmp['type'] = string.format("%s",class_type)
-              end
-              if (types.struct[class_type]) then
-                tmp['class']= "struct"
-                tmp['type'] = string.format("%s",class_type)
-              end
-
-              tmp['mandatory'] = v2:attr('mandatory') or 'true'
-              if (v2:attr('array')) then
-                tmp['array'] = v2:attr('array')
-              end
-              if (v2:attr('minsize') and v2:attr('maxsize')) then
-                tmp['minsize'] = v2:attr('minsize')
-                tmp['maxsize'] = v2:attr('maxsize')
-              end
-              if (v2:attr('minlength')) then tmp['minlength'] = tonumber(v2:attr('minlength')) end
-              if (v2:attr('maxlength')) then tmp['maxlength'] = tonumber(v2:attr('maxlength')) end
-              if (v2:attr('minvalue')) then tmp['minvalue'] = tonumber(v2:attr('minvalue')) end
-              if (v2:attr('maxvalue')) then tmp['maxvalue'] = tonumber(v2:attr('maxvalue')) end
-
-              retval[v2:attr('name')] = tmp
-            end
-          end
-        end
+  for k1,v1 in pairs(func_schema) do
+      if (type(user_data)~='table')then
+        return false, "not valid type of "..k1 .. " expected structure, got "..type(user_data)
       end
+      local bool_result = true
+      if((user_data[k1]==nil) and func_schema[k1]["mandatory"] == true) then
+        bool_result = false
+        error_message["mandatory "..k1]= "mandatory parameter "..k1.." not present"
+      end
+      result = bool_result and result
     end
+  return result, error_message
+end
 
-    return retval
+local function GetNames( name )
+  local interface,  short_name
+  if (string.find(name, "%.")) then
+    interface, short_name = name:match("([^.]+).([^.]+)")
+  else
+    interface ='Ford Sync RAPI'
+    short_name = name
   end
+  return interface, short_name
+end
 
-  local function errorMsgToString(tbl)
-    local tmp = ''
-    if type(tbl) == 'table' then
-      for _,v in pairs(tbl) do
-        if type(v) == 'table' then
-          tmp = tmp ..errorMsgToString(v) ..'\n'
-        else
-          tmp = tmp ..v ..'\n'
-        end
-      end
-      return tmp
-    end
+function module.mt.__index:CompareStructs( data_elem, struct_schema )
+  local result = true
+  local result1 = true --for mandatory param
+  local result2 = true -- for correct types of param
```

Comment:
May be better to rename to the functionality of this variables?
like `MandatoryCheckResult' and `AprametersTypeCheckResult`